llvm/lib/Transforms/Vectorize/VPlan.cpp

```diff
@@ -1266,6 +1293,15 @@ VPlan *VPlan::duplicate() {
   return NewPlan;
 }
 
+VPBasicBlock *VPlan::getScalarPreheader() {
+  auto *MiddleVPBB =
+      cast<VPBasicBlock>(getVectorLoopRegion()->getSingleSuccessor());
+  auto *LastSucc = MiddleVPBB->getSuccessors().back();
+  // Order is strict: if the last successor is VPIRBasicBlock, it must be the
+  // single exit.
```

Comment:
```suggestion
  // If scalar preheader is connected to VPlan, it is the last successor of MiddleVPBB.
  // If this last successor is a VPIRBasicBlock, it is the Exit block rather than the scalar preheader.
```